text/0000-api-evolution.md

```diff
@@ -0,0 +1,748 @@
+
+**Globs with fine-grained control**
+
+Another useful tool for working with globs would be the ability to *exclude*
+certain items from a glob import, e.g. something like:
+
+```rust
+use some_module::{* without Foo};
+```
+
+This is especially useful for the case where multiple modules being glob
+imported happen to export items with the same name.
+
+**"Extensible" enums**
+
+There is already [an RFC](https://github.com/rust-lang/rfcs/pull/757) for an
+`enum` annotation that would make it possible to add variants without ever
+breaking downstream code.
+
+**Sealed traits**
+
+The ability to annotate a trait with some "sealed" marker, saying that no
+external implementations are allowed, would be useful in certain cases where a
+crate wishes to define a closed set of types that implements a particular
+interface. Such an attribute would make it possible to evolve the interface
+without a major version bump (since no downstream implementors can exist).
+
+**Defaulted parameters**
+
+Also known as "optional arguments" -- an
+[oft-requested](https://github.com/rust-lang/rfcs/issues/323) feature. Allowing
+arguments to a function to be optional makes it possible to add new arguments
+after the fact without a major version bump.
+
+# Drawbacks and Alternatives
+
+The main drawback to the approach laid out here is that it makes the stability
+and semver guarantees a bit fuzzier: the promise is not that code will never
+break, full stop, but rather that minor release breakage is of an extremely
+limited form, for which there are a variety of mitigation strategies. This
+approach tries to strike a middle ground between a very hard line for stability
+(which, for Rust, would rule out many forms of extension) and willy-nilly
+breakage: it's an explicit, but pragmatic policy.
+
+An alternative would be to take a harder line and find some other way to allow
+API evolution. Supposing that we resolved the issues around glob imports, the
+main problems with breakage have to do with adding new inherent methods or trait
+implementations -- both of which are vital forms of evolution. It might be
+possible, in the standard library case, to provide some kind of version-based
+opt in to this evolution: a crate could opt in to breaking changes for a
+particular version of Rust, which might in turn be provided only through some
+`cfg`-like mechanism.
+
+Note that these strategies are not mutually exclusive. Rust's development
+processes involved a very steady, strong stream of breakage, and while we need
+to be very serious about stabilization, it is possible to take an iterative
+approach. The changes considered "major" by this RFC already move the bar *very
+significantly* from what was permitted pre-1.0. It may turn out that even the
+minor forms of breakage permitted here are, in the long run, too much to
+tolerate; at that point we could revise the policies here and explore some
+opt-in scheme, for example.
+
+# Unresolved questions
+
+## Behavioral issues
+
```

Comment:
Reading the first item below, I wondered: what does adding a `#[must_use]` annotation to a type qualify as?  Seems like a major change to me -- or is that lint only warn-by-default?

And are there other attributes in this bucket, where adding them should be considered a major change?  (E.g. changes to `#[inline]` directives are probably safe, but then there's stuff like `#[unsafe_no_drop_flag]` ... though  I guess that is unstable anyway, right?)